chain/acala/acala.go

```diff
@@ -0,0 +1,169 @@
+// via the Go-substrate RPC API.
+type Client struct {
+	opts ClientOptions
+	api  gsrpc.SubstrateAPI
+}
+
+// NewClient constructs a new Acala client using the client options.
+func NewClient(opts ClientOptions) (*Client, error) {
+	substrateAPI, err := gsrpc.NewSubstrateAPI(string(opts.rpcURL))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		opts: opts,
+		api:  *substrateAPI,
+	}, nil
+}
+
+// BurnLogInput defines a structure used to serialize relevant data to fetch
+// Acala burn logs from RenVmBridge's system events.
+type BurnLogInput struct {
+	Blockhash pack.Bytes32
+	ExtSign   pack.Bytes
+}
+
+// BurnLogOutput defines a structure used to represent burn data from Acala.
+type BurnLogOutput struct {
+	Amount    pack.U256
+	Recipient address.RawAddress
+	Confs     pack.U64
+}
+
+// CallContractSystemEvents fetches burn logs from the substrate system events.
+func (client *Client) CallContractSystemEvents(_ context.Context, _ address.Address, calldata contract.CallData) (pack.Bytes, error) {
+	// Deserialise the calldata bytes.
+	input := BurnLogInput{}
+	if err := surge.FromBinary(&input, calldata); err != nil {
+		return pack.Bytes{}, fmt.Errorf("deserialise calldata: %v", err)
+	}
+
+	// Get chain metadata.
+	meta, err := client.api.RPC.State.GetMetadataLatest()
+	if err != nil {
+		return pack.Bytes{}, fmt.Errorf("get metadata: %v", err)
+	}
+
+	// This key is used to read the state storage at the block of interest.
+	key, err := types.CreateStorageKey(meta, "System", "Events", nil, nil)
+	if err != nil {
+		return pack.Bytes{}, fmt.Errorf("create storage key: %v", err)
+	}
+
+	// Get the block in which the burn event was logged.
+	block, err := client.api.RPC.Chain.GetBlock(types.Hash(input.Blockhash))
+	if err != nil {
+		return pack.Bytes{}, fmt.Errorf("get block: %v", err)
+	}
+
+	// Get the latest block header. This will be used to calculate number of block
+	// confirmations of the burn log of interest.
+	header, err := client.api.RPC.Chain.GetHeaderLatest()
+	if err != nil {
+		return pack.Bytes{}, fmt.Errorf("get header: %v", err)
+	}
+
+	// Retrieve raw bytes from storage at the block and storage key of interest.
+	data, err := client.api.RPC.State.GetStorageRaw(key, types.Hash(input.Blockhash))
+	if err != nil {
+		return pack.Bytes{}, fmt.Errorf("get storage: %v", err)
+	}
+
+	// Fetch the extrinsic's index in the block.
+	extID := -1
+	for i, ext := range block.Block.Extrinsics {
+		if input.ExtSign.Equal(pack.Bytes(ext.Signature.Signature.AsSr25519[:])) {
+			extID = i
+			break
+		}
+	}
+	if extID == -1 {
+		return pack.Bytes{}, fmt.Errorf("extrinsic not found in block")
+	}
+
+	// Decode the event data to get the burn log.
+	burnEvent, err := decodeEventData(meta, data, uint32(extID))
+	if err != nil {
+		return pack.Bytes{}, err
+	}
+
+	// Calculate block confirmations for the event.
+	confs := header.Number - block.Block.Header.Number + 1
```

Comment:
If the latest block header is the same as that of the event, won't that result in `-1` here causing an overflow when converting to a uint64 later on?